### PR TITLE
create TRX from LazyTractogram or Generator

### PR DESCRIPTION
**Pull request by frheault**
Create TRX file from Generator or LazyTractogram and write to memmap in chunks.
Support a buffer to avoid resizing the TRX at every chunk.

Added tests to verify if it is working.

### CODE REVIEW

**Comment by frheault** [comment]
@arokem It is still a bit rought, but is this what you had in mind? For me it is working pretty well.

In the `tracking_introduction_eudx.py` example from Dipy doing this works:
```
streamlines_generator = LocalTracking(csa_peaks, stopping_criterion, seeds,
                                      affine=affine, step_size=.5)
#streamlines = Streamlines(streamlines_generator)
trx = TrxFile.from_lazy_tractogram(streamlines_generator, hardi_img, extra_buffer=100000)
```

Using an extra_buffer of 0 will recreate a new TrxFile at each chunk (few thousand streamlines) which is a lot of new temporary folders and all and rewriting data.
Using a bigger extra_buffer will create an empty TrxFile that can use preallocations and only get resized (if needed) a few times and the function finishes with a final resize.

I did not benchmark it.

File: trx/tests/test_memmap.py [hunk 1]
```diff
@@ -178,12 +174,117 @@ def test_load_directory(path):
...
+
+
+@pytest.mark.parametrize("path, size, buffer", [("small.trx", 50, 10000),
+                                                ("small.trx", 0, 10000),
+                                                ("small.trx", 25000, 10000),
+                                                ("small.trx", 50, 0),
+                                                ("small.trx", 0, 0),
+                                                ("small.trx", 25000, 10000)])
+def test_from_lazy_tractogram(path, size, buffer):
+    DATA = {}
```
Comment (arokem) [inline trx/tests/test_memmap.py:250]: I am a bit confused: why are you setting up this dict, rather than assigning these as variables?

**Review by arokem — COMMENTED** [review]
Yeah - overall looks great! Might be worth adding the example you have in the message on the PR to the documentation?

**Comment by arokem** [comment]
Is this still WIP?

**Comment by frheault** [comment]
> Is this still WIP?

With #37 passed and the test passing, I think this is ready for testing.

**Review by arokem — APPROVED** [review]

**Comment by frheault** [comment]
I need to test with larger files to quantify how much time is saved or how much allocations are avoided.

I know it works but I want to be able to have a few numbers to back it up.

**Comment by arokem** [comment]
Sounds good! No rush.

**Pull request merged**